test/runnable/ufcs.d

```diff
@@ -768,26 +768,6 @@ void test10041()
     assert(writeln10041(aa) == "int[int]");
 }
 
-/*******************************************/
-// https://issues.dlang.org/show_bug.cgi?id=10047
-
-struct Typedef10047(T)
-{
-    template opDispatch(string name)
-    {
-        static assert(0);
-    }
-}
-
-struct A10047 {}
-int foo10047(Typedef10047!A10047 a) { return 10; }
-
-void test10047()
-{
-    Typedef10047!A10047 a;
-    assert(a.foo10047() == 10);
```

Comment:
I am confused by this test. Should it not fail when opDispatch is instantiated? It seems to me that the static assert should fail at compile time when it is encountered, but indeed, opDispatch is instantiated eagerly in this case since the free function is a better match.

Reply:
It seems like opDispatch should be considered only after ufcs has been tried.

Reply:
no, opDispatch should be considered before ufcs. Members always go before ufcs.

Reply:
so this breaking change should be fine then?

Reply:
yeah the code in that test should just fail with an error message. If you don't want opDispatch considered, use a constraint to remove it from consideration. I think this misunderstanding in the original PR is what created this whole mess.

Reply:
Hmm, so if you define `opDispatch` for a struct then you automatically disable (or hijack?) ufcs. Is this something we want? I agree that properties should take precedence over ufcs, however, since everyone loves ufcs it is kind of problematic to disable it once an `opDispatch` is defined. Maybe it would be easier to simply let ufcs take precedence over `opDispatch` if there is a function that matches. This way we won't break any code out there. Note that phobos also uses `this.move` inside a struct that defines `opDispatch` => this is what is causing the current phobos failure: suddenly opDispatch is preferred to ufcs.

Reply:
From my perspective, `opDispatch` is a last resort that should be taken only when all other possibilities have been exhausted.

Reply:
Unfortunately, given the current code, it is impossible to know whether opDispatch failed because errors in the body or because the definition does not match the call. Implementing this will require ugly code to be added. I am starting to wonder if it wouldn't be better to put this fix behind a transition flag and add a changelog entry.

Reply:
The spec says:
> Member names not found in a class or struct can be forwarded to a template function named opDispatch for resolution. 

and

>     // If `T` does not have member function `sun`,
>    // `t.sun(1)` is interpreted as if it were written `sun(t, 1)`

opDispatch is a member function for looking up member functions. Member functions are ALWAYS checked BEFORE ufcs - the spec says ufcs is considered only "if `T` does not have member function".

If the code cannot support this, the code must be fixed. Matching a template is supposed to be distinct from compiling a template and UFCS is not permitted to hijack member functions and the current behavior cause real damage.

Reply:
Practically speaking, opDispatch is *part of the type*, UFCS is outside the type. You have layers that apply. It should go:

1. standard members
2. opDispatch
3. alias this
4. UFCS

If any layer matches, it should be preferred over the subsequent layers. This way the type's author is in control.

Really, this status quo does not change with a fix for this issue -- it's just that you now need to be more careful about opDispatch constraints.

> Unfortunately, given the current code, it is impossible to know whether opDispatch failed because of errors in the body or because the definition does not match the call.

Ugh, confirmed in a test. That really stinks. I think this is probably a reason why this hasn't been fixed yet (though attempts have been made).

I don't think we can release this if we don't have a deprecation path for existing code that relies on the old behavior.

Reply:
We could always put it behind a transition switch which in time will become the default.

Reply:
Can we make a 3-way transition switch? i.e. 

mode 1: current behavior
mode 2: opDispatch fails, no UFCS available, you get opDispatch failure messages. if UFCS is available, then calls UFCS, BUT you get a deprecation warning (to help you find these cases)
mode 3: your PR as written -- an opDispatch failure is a hard failure if it's selected, even if UFCS is available.

That way, one can try and ferret out all the places they have unconstrained opDispatch.

Reply:
Actually, mode 3 should find those too, since it should fail to compile.